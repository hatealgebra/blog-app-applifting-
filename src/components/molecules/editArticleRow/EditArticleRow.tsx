import React from "react";

import Checkbox from "../../atoms/checkbox/Checkbox";
import {
  StyledEditArticleRow,
  StyledIconButton,
} from "./editArticleRow.styled";

import { FiEdit2 } from "@react-icons/all-files/fi/FiEdit2";
import { RiDeleteBin7Line } from "@react-icons/all-files/ri/RiDeleteBin7Line";

const EditArticleRow = ({
  iteration,
  articleId,
  title,
  perex,
  comments,
  deleteArticle,
  editArticle,
  ...props
}: EditArticleRowProps) => {
  const [isChecked, setIsChecked] = React.useState(false);

  React.useEffect(() => {
    props.setCheckedBoxes((prevState) => {
      const newState = prevState || {};
      newState[iteration] = isChecked;
      return newState;
    });
    return () => {};
  }, [isChecked]);

  React.useEffect(() => {
    setIsChecked(props.isChecked);
  }, [props.isChecked]);

  return (
    <StyledEditArticleRow
      data-testid="edit-article-row"
      className="edit-article"
    >
      <td className="edit-article__checkbox">
        <Checkbox
          isChecked={isChecked}
          onChange={() => setIsChecked((prev) => !prev)}
        />
      </td>
      <td className="edit-article__title">{title}</td>
      <td className="edit-article__perex">{perex}</td>
      <td className="edit-article__author">author</td>
      <td className="edit-article__nr-comments">{comments}</td>
      <td className="edit-article__actions-container">
        <div className="edit-article__actions">
          <StyledIconButton
            className="edit-article__button-edit"
            onClick={editArticle}
          >
            <FiEdit2 size="1.2em" />
          </StyledIconButton>
          <StyledIconButton
            className="edit-article__button-delete"
            onClick={() => deleteArticle(articleId!)}
          >
            <RiDeleteBin7Line size="1.2em" />
          </StyledIconButton>
        </div>
      </td>
    </StyledEditArticleRow>
  );
};

interface EditArticleRowProps {
  iteration: number;
  articleId?: string;
  title?: string;
  perex?: string;
  comments?: number;
  deleteArticle: Function;
  editArticle: React.MouseEventHandler<HTMLButtonElement>;
  isChecked: boolean;
  setCheckedBoxes: React.SetStateAction<any>;
}

export default EditArticleRow;
