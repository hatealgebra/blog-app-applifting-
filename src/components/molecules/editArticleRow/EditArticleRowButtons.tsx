import React, { Dispatch } from 'react';
import { AnyAction } from '@reduxjs/toolkit';
import { components } from '../../../customTypes/declarations';
import Checkbox from '../../atoms/checkbox/Checkbox';
import { StyledEditArticleRow } from './editArticleRow.styled';

import ButtonSort from '../../atoms/button/ButtonSort';
import { sortMyArticles } from '../../../store/slices/admin.slices';
import { deleteArticleThunk } from '../../../store/thunks/admin.thunks';

interface EditArticleRowButtonsProp {
  originalArray: components['schemas']['Article'];
  switchAllBoxes: React.SetStateAction<any>;
  dispatch: React.Dispatch<AnyAction>;
}

const EditArticleRowButtons = ({
  originalArray,
  switchAllBoxes,
  dispatch,
}: EditArticleRowButtonsProp) => {
  const { BY_TITLE, BY_AUTHOR, BY_NR_COMMENTS, BY_PEREX, ORIGINAL } =
    ESortByOptions;
  const [isChecked, setIsChecked] = React.useState(false);
  const [isActive, setIsActive] = React.useState<ESortByOptions>(ORIGINAL);

  const activeSort = (sortType: ESortByOptions) => {
    setIsActive((prev) => {
      if (prev === sortType) {
        dispatch(sortMyArticles(ORIGINAL));
        return ORIGINAL;
      }
      dispatch(sortMyArticles(sortType));
      return sortType;
    });
  };

  React.useEffect(() => {
    switchAllBoxes(isChecked);
  }, [isChecked]);

  React.useEffect(() => {
    dispatch({ type: 'admin/sortMyArticles', payload: isActive });
  }, [originalArray]);

  return (
    <StyledEditArticleRow className="edit-article">
      <th className="edit-article__checkbox">
        <Checkbox
          isChecked={isChecked}
          onChange={() => setIsChecked((prev) => !prev)}
        />
      </th>
      <th className="edit-article__title">
        <ButtonSort
          isActive={isActive === BY_TITLE}
          onClick={() => activeSort(BY_TITLE)}
        >
          Article title
        </ButtonSort>
      </th>
      <th className="edit-article__perex">
        <ButtonSort
          isActive={isActive === BY_PEREX}
          onClick={() => activeSort(BY_PEREX)}
        >
          Perex
        </ButtonSort>
      </th>
      <th className="edit-article__author">
        <ButtonSort
          isActive={isActive === BY_AUTHOR}
          onClick={() => activeSort(BY_AUTHOR)}
        >
          Author
        </ButtonSort>
      </th>
      <th className="edit-article__nr-comments">
        <ButtonSort
          isActive={isActive === BY_NR_COMMENTS}
          onClick={() => activeSort(BY_NR_COMMENTS)}
        >
          # comments
        </ButtonSort>
      </th>
      <th className="edit-article__actions-container">Actions</th>
    </StyledEditArticleRow>
  );
};

export enum ESortByOptions {
  ORIGINAL = 'original',
  BY_TITLE = 'title',
  BY_PEREX = 'perex',
  BY_AUTHOR = 'author',
  BY_NR_COMMENTS = 'comments',
}

export default EditArticleRowButtons;
