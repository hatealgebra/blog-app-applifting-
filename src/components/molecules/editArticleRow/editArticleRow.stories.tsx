import React from "react";

import EditArticleRow from "./EditArticleRow";
import EditArticleRowButtons from "./EditArticleRowButtons";

import articlesMockJSON from "../../../__mocks__/asyncData/get/allArticlesResponse.mock.json";
import { action } from "@storybook/addon-actions";

const { title, perex, comments } = articlesMockJSON.items[0];

export const ArticleRowExample = () => {
  const editArticleMock = () => action("Edit function called");
  const callDeleteMock = () => action("Delete article right");
  const setCheckedBoxes = () => action("All boxes were checked");

  return (
    <EditArticleRow
      iteration={0}
      title={title}
      perex={perex}
      comments={comments}
      editArticle={editArticleMock}
      deleteArticle={callDeleteMock}
      isChecked={true}
      setCheckedBoxes={setCheckedBoxes}
    />
  );
};

export const ArticleRowButtonsExample = () => {
  return (
    <EditArticleRowButtons
      setCheckAll={action("setCheckAll called!")}
      dispatchSort={action("Dispatch sort called!")}
      dispatchEdit={action("Dispatch edit called")}
      dispatchDelete={action("Dispatch delete called!")}
      dispatchReset={action("Dispatch reset called!")}
    />
  );
};

export default {
  title: "Molecules/Edit article",
  component: EditArticleRow,
};
