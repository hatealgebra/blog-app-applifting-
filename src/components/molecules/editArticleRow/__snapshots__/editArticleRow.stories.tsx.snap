// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molecules/Edit article ArticleRowButtonsExample smoke-test 1`] = `""`;

exports[`Molecules/Edit article ArticleRowExample smoke-test 1`] = `
<tr data-testid="edit-article-row"
    class="sc-jrkPcm jbOSku edit-article"
>
  <td class="edit-article__checkbox">
    <div class="sc-ftLKQv iibJaj">
      <input type="checkbox"
             class="sc-iBAaJG ciwLrW"
      >
    </div>
  </td>
  <td class="edit-article__title">
    Wet vs. Dry Cat Food: Which is Better?
  </td>
  <td class="edit-article__perex">
    A cat's whiskers — or vibrissae — are a well-honed sensory tool that helps a cat see in the dark and steer clear of hungry predators. Whiskers are highly ...
  </td>
  <td class="edit-article__author">
    author
  </td>
  <td class="edit-article__nr-comments">
  </td>
  <td class="edit-article__actions-container">
    <div class="edit-article__actions">
      <button class="sc-pqitP fWIcxZ edit-article__button-edit">
        <svg stroke="currentColor"
             fill="none"
             stroke-width="2"
             viewbox="0 0 24 24"
             stroke-linecap="round"
             stroke-linejoin="round"
             height="1.2em"
             width="1.2em"
             xmlns="http://www.w3.org/2000/svg"
        >
          <path d="M17 3a2.828 2.828 0 1 1 4 4L7.5 20.5 2 22l1.5-5.5L17 3z">
          </path>
        </svg>
      </button>
      <button class="sc-pqitP fWIcxZ edit-article__button-delete">
        <svg stroke="currentColor"
             fill="currentColor"
             stroke-width="0"
             viewbox="0 0 24 24"
             height="1.2em"
             width="1.2em"
             xmlns="http://www.w3.org/2000/svg"
        >
          <g>
            <path fill="none"
                  d="M0 0h24v24H0z"
            >
            </path>
            <path d="M17 6h5v2h-2v13a1 1 0 0 1-1 1H5a1 1 0 0 1-1-1V8H2V6h5V3a1 1 0 0 1 1-1h8a1 1 0 0 1 1 1v3zm1 2H6v12h12V8zM9 4v2h6V4H9z">
            </path>
          </g>
        </svg>
      </button>
    </div>
  </td>
</tr>
`;
